extensions/microsoft-authentication/src/node/authProvider.ts

```diff
@@ -187,86 +186,70 @@ export class MsalAuthProvider implements AuthenticationProvider {
 
 		this._logger.info('[createSession]', `[${scopeData.scopeStr}]`, 'starting');
 		const cachedPca = await this.getOrCreatePublicClientApplication(scopeData.clientId, scopeData.tenant);
-		let result: AuthenticationResult | undefined;
-
-		try {
-			const windowHandle = window.nativeHandle ? Buffer.from(window.nativeHandle) : undefined;
-			result = await cachedPca.acquireTokenInteractive({
-				openBrowser: async (url: string) => { await env.openExternal(Uri.parse(url)); },
-				scopes: scopeData.scopesToSend,
-				// The logic for rendering one or the other of these templates is in the
-				// template itself, so we pass the same one for both.
-				successTemplate: loopbackTemplate,
-				errorTemplate: loopbackTemplate,
-				// Pass the label of the account to the login hint so that we prefer signing in to that account
-				loginHint: options.account?.label,
-				// If we aren't logging in to a specific account, then we can use the prompt to make sure they get
-				// the option to choose a different account.
-				prompt: options.account?.label ? undefined : 'select_account',
-				windowHandle
-			});
-		} catch (e) {
-			if (e instanceof CancellationError) {
-				const yes = l10n.t('Yes');
-				const result = await window.showErrorMessage(
-					l10n.t('Having trouble logging in?'),
-					{
-						modal: true,
-						detail: l10n.t('Would you like to try a different way to sign in to your Microsoft account? ({0})', 'protocol handler')
-					},
-					yes
-				);
-				if (!result) {
-					this._telemetryReporter.sendLoginFailedEvent();
-					throw e;
-				}
+
+		// Used for showing a friendlier message to the user when the explicitly cancel a flow.
+		let userCancelled: boolean | undefined;
+		const yes = l10n.t('Yes');
+		const no = l10n.t('No');
+		const promptToContinue = async (mode: string) => {
+			if (userCancelled === undefined) {
+				// We haven't had a failure yet so wait to prompt
+				return;
 			}
-			// This error comes from the backend and is likely not due to the user's machine
-			// failing to open a port or something local that would require us to try the
-			// URL handler loopback client.
-			if (e instanceof ServerError) {
-				this._telemetryReporter.sendLoginFailedEvent();
-				throw e;
+			const message = userCancelled
+				? l10n.t('Having trouble logging in? Would you like to try a different way? ({0})', mode)
+				: l10n.t('You have not yet finished authorizing this extension to use your Microsoft Account. Would you like to try a different way? ({0})', mode);
+			const result = await window.showWarningMessage(message, yes, no);
+			if (result !== yes) {
+				throw new CancellationError();
 			}
+		};
 
-			// The user closed the modal window
-			if ((e as ClientAuthError).errorCode === ClientAuthErrorCodes.userCanceled) {
-				this._telemetryReporter.sendLoginFailedEvent();
-				throw e;
-			}
+		const flows = getMsalFlows({
+			extensionHost: typeof navigator === 'undefined'
```

Comment:
💭 Nested ternary expressions a bit hard to read especially when inside the other expression, would it be better to have a `extensionHost` variable for it declared above?